docs/feat: zk lightnet (#780)
o1-labs/docs2: docs/zkapps/lightnet.mdx

---
title: Lightnet zkApp Testing
hide_title: true
description: Use lightweight Mina blockchain network (Lightnet) for testing on a local blockchain before you test with a live network. 
keywords:
  - zkApp CLI
  - lightnet
  - zkApp testing
  - local testing
  - using zk lightnet
---

# Lightnet zkApp Testing

The most efficient way to deploy and run a lightweight Mina blockchain network for testing purposes is to spin up a lightweight Mina network (Lightnet) in a single Docker container. Lightnet is a resource-efficient solution with fast startup and syncing times.

Lightnet lets you test your zkApp locally on an accurate representation of Mina blockchain before you test with a live network.

> Using Lightnet provides the following benefits:

Reviewer: ```suggestion
Lightnet provides the following benefits:
```

- Reduces the time from ideation to launch by letting you test your applications and changes against a close-to-real Mina network.
- Starts a single-node network or a multi-node network with diverse types of participants.
- Creates and funds accounts so that you can deploy and interact with your zkApp. The genesis ledger is configured with 1000 pre-funded accounts with a 1550 MINA balance on each account.

> - Optionally runs the Mina [archive node](/node-operators/archive-node).

Reviewer: ```suggestion
- Optionally runs a Mina [archive node](/node-operators/archive-node).
```

- Provides the Mina accounts manager helper tool so you can automate account retrieval.
- Launches the lightweight Mina Explorer in a web browser so you can monitor transactions.

> 

Reviewer: ```suggestion
- Simplifies zkApp troubleshooting by providing convenient access to detailed logs.
```

:::info

> Use of Lightnet blockchain is appropriate for local development and testing only. It is not intended to replicate all aspects of the full network.

Reviewer: ```suggestion
Use of Lightnet blockchain is appropriate for local development and testing only. It is not intended to replicate all aspects of the full network.
```
```suggestion
Use of Lightnet is appropriate for local development and testing only. It is not intended to replicate all aspects of the full network.
```

:::

## Prerequisites

Docker Engine is required for your environment. 

1. [Install Docker Engine](
https://docs.docker.com/engine/install/).
1. Start Docker Engine. 

Docker Engine must be running before you can start the Docker container.

## Start a single node network

To start a single node network with default behavior:

```sh
zk lightnet start
```

> This command uses default settings that successfully serve the majority of testing requirements:

Reviewer 2: ```suggestion
This command performs the following operations using default settings that successfully serve the majority of testing requirements:
```
Could probably be better still. The original version was implying that what follows is a list of testing requirements but it's actually a list of things the command does. Trying to reflect that more clearly

To keep your current working versions of o1js and Mina:

```sh
zk lightnet start --no-pull
```

## Stop the local network

To stop the local network, remove the Docker container, and clean up the environment:

```sh
zk lightnet stop
```

When Lightnet is stopped, the log files for Docker container services are saved to the host file system at `/Users/<username>/.cache/zkapp-cli/lightnet/logs/`.

To disable saving of log files:

```sh
zk lightnet stop --no-save-logs
```

## Troubleshooting transactions

Tools are provided for troubleshooting.

### Lightweight Mina Explorer

To monitor transactions, launch the lightweight Mina Explorer:

```sh
zk lightnet explorer
```

By default, the latest version is launched. 

To list versions, their published dates, and show the version in use:

```sh
zk lightnet explorer list true
```

To use a specific version of the lightweight Mina Explorer:

```sh
zk lightnet explorer use <version>
```

> ### Reinstantiate lightnet

Reviewer 2: Section missing?

Author: there is no instantiate command, so I have removed. Love this eagle eye feedback, thank you

### Log files

> Log files for different processes are saved inside the Docker container as:

Reviewer: ```suggestion
Log files for various processes are saved inside the Docker container as:
```

- `/root/logs/*.log`
- `/root/.mina-network/mina-local-network-2-1-1/nodes/**/logs/*.log`

To save the log files that are produced by Docker container processes to the host machine file system:

```sh
zk lightnet logs save
```

To debug your zkApp, you can follow the Docker container process logs in real time. 

On a different terminal window, to stream the process logs:

```sh
zk lightnet logs follow
```

Select the Docker container process to follow. Press Ctrl+C to stop streaming.

## Lightnet status

To get the network status:

```sh
zk lightnet status
```

> The network status is returned, including useful URLs. For example, the default single node network status returns these useful URLs:

Reviewer 2: ```suggestion
The network status is returned, including useful URLs. For example, the default single node network status returns these endpoints:
```
??

Author: Your question marks mean you have a question? Can you elaborate please? @nicc

**Advanced** To access the genesis ledger configuration file `daemon.json`, you must manually access the Docker container file system with `docker exec -it mina-local-lightnet bash`. 

### zkApp Account Interactions

The `acquireKeyPair()`, `releaseKeyPair()`,  and `listAcquiredKeyPairs()` methods in the `Lightnet` o1js API namespace handle zkApp access to account information, including public keys, private keys, and secret keys. For details, see the [doc comments](https://github.com/o1-labs/o1js/blob/f5db874a3fd52635d3d77628696bc8e97f78b76c/src/lib/fetch.ts#L1012) in the code. 

For a real-world example of using Lightnet accounts, see [run_live.ts](https://github.com/o1-labs/o1js/blob/main/src/examples/zkapps/hello_world/run_live.ts) example file used in [Tutorial 1: Hello World](/zkapps/tutorials/hello-world).

### Account Use Other Than zkApp

To get the accounts information outside of the zkApp program, use the Accounts-Manager Docker container service at the endpoint: http://localhost:8181/. This endpoint is the same for all users and is available when the Lightnet Docker container is up and running.

> <!-- 

Reviewer 2: should be removed before merging

> Question for Serhii: how to release accounts? 

Author: ```suggestion
```

Reviewer 3: https://github.com/o1-labs/docs2/pull/780#discussion_r1465175450

**Advanced** To access the genesis ledger configuration file `daemon.json`, you must manually access the Docker container file system with `docker exec -it mina-local-lightnet bash`. 

### zkApp Account Interactions

The `acquireKeyPair()`, `releaseKeyPair()`,  and `listAcquiredKeyPairs()` methods in the `Lightnet` o1js API namespace handle zkApp access to account information, including public keys, private keys, and secret keys. For details, see the [doc comments](https://github.com/o1-labs/o1js/blob/f5db874a3fd52635d3d77628696bc8e97f78b76c/src/lib/fetch.ts#L1012) in the code. 

For a real-world example of using Lightnet accounts, see [run_live.ts](https://github.com/o1-labs/o1js/blob/main/src/examples/zkapps/hello_world/run_live.ts) example file used in [Tutorial 1: Hello World](/zkapps/tutorials/hello-world).

### Account Use Other Than zkApp

To get the accounts information outside of the zkApp program, use the Accounts-Manager Docker container service at the endpoint: http://localhost:8181/. This endpoint is the same for all users and is available when the Lightnet Docker container is up and running.

> <!-- 

Author: @Sventimir  how do we release the unused accounts?

Author: I will benefit from your expertise and pairing to work through the accounts

Reviewer 3: If we are talking not about the o1js API then you can refer to https://github.com/shimkiv/mina-accounts-manager readme where available HTTP endpoints presented.
`GET http://localhost:8181/acquire-account`
`PUT http://localhost:8181/release-account (with json body)`
